### PR TITLE
Added more detailed documentation plus several code examples for ImportString (Take 2)

### LINKED ISSUE #537: Add Import type
# Feature Request

Would there be interest in adding an `Import` type so you could do things like:

```
class AppConfig(BaseModel):
    middleware: List[Import] = []

config = AppConfig(middlewear=['diary.api:make_session'])

...

config.make_session()
```

I have a prototype [here](https://github.com/cjw296/diary/blob/dcfe5da44bef834a9780696393fb8ca000baeb1a/backend/diary/config.py#L11-L29) and I'm happy to work it up into a PR, but thought I'd check if there was interest first.

### PR DESCRIPTION
**Pull request by ybressler**
> Previous PR (#4893) got gnarly. Created a new branch and cherry picked commits.

Added more detailed documentation plus several code examples for `ImportString`.

### Some important details:
* Created a new subsection in `types` docs for "exotic types"
  * The term "exotic type" is pretty fun, and I have seen it used loosely in various documentation guides. But I don't have a "text book definition" for it. What I mean to say is, perhaps `"exotic types"` is appropriate section title, perhaps not.
* Added some examples of expected "good behavior" for `ImportString`
* Added some examples of expected "bad behavior"  for `ImportString`
* Added an example of _**unexpected "bad behavior"**_ as well 😢
  * Maybe there's a solution we can employ (or a workaround) in the code example. If not, I think it's best to document what doesn't work, save people some headaches.

**Side note:** This is my first PR for pydantic! Woot woot!

fix #537

### CODE REVIEW

File: docs/usage/types.md [hunk 1]
```diff
@@ -584,6 +585,28 @@ With proper ordering in an annotated `Union`, you can use this to parse types of
 : type method for constraining strs;
   see [Constrained Types](#constrained-types)
 
+### Exotic Types
+Pydantic also supports exotic types, which are described in detail below:
+
+#### ImportString
+On model instantiation, pointers will be evaluated and imported,
+(distantly related behavior to [PyObject](https://docs.python.org/3/c-api/structures.html#c.PyObject)).
```
Comment (samuelcolvin) [inline docs/usage/types.md:593]: I'm not sure this line really helps with the explanation.

File: docs/usage/types.md [hunk 2]
```diff
@@ -910,7 +933,7 @@ to get validators to parse and validate the input data.
 
 !!! tip
     These validators have the same semantics as in [Validators](validators.md), you can
-    declare a parameter `config`, `field`, etc.
+    declare a parameter `config`,types_import_string_1.py `field`, etc.
```
Comment (samuelcolvin) [inline docs/usage/types.md:936]:
```suggestion
    declare a parameter `config`, `field`, etc.
```

File: docs/usage/types.md [hunk 3]
```diff
@@ -584,6 +585,28 @@ With proper ordering in an annotated `Union`, you can use this to parse types of
...
+Pydantic also supports exotic types, which are described in detail below:
+
+#### ImportString
+On model instantiation, pointers will be evaluated and imported,
+(distantly related behavior to [PyObject](https://docs.python.org/3/c-api/structures.html#c.PyObject)).
+There is some nuance to this behavior, demonstrated in the examples below.
+
+**Good behavior:**
+{!.tmp_examples/types_import_string_1.md!}
+(This script is complete, it should run "as is")
```
Comment (hramezani) [inline docs/usage/types.md:598]: Not sure do we need this line here or not. seems the `exec_examples` will add this line:

https://github.com/pydantic/pydantic/blob/dcf35678764be812048fabd4b8f4c577f5c6e267/docs/build/exec_examples.py#L391

**Comment by samuelcolvin** [comment]
please update.

File: docs/examples/types_import_string_3.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+
+# So now that we know that after an ImportString is evaluated by Pydantic
+# it results in its underlying object, we can configure our json encoder
+# to account for those specific types
+class WithCustomEncodersGood(BaseModel):
+    obj: ImportString
+
+    class Config:
+        json_encoders = {
```
[the remark this answers is not in the document]
Comment (ybressler) [reply docs/examples/types_import_string_3.py:36]: Good point. I'll open a follow up issue for this. Bonus for development work >> I'm familiar with the use case, so implementing won't be too challenging.

File: docs/usage/types.md (hunk 1 above)
Comment (ybressler) [reply docs/usage/types.md:593]: Removed the link.

File: docs/examples/types_import_string_2.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+from pydantic import BaseModel, ImportString
```
[the remark this answers is not in the document]
Comment (ybressler) [reply docs/examples/types_import_string_2.py:1]: Here's the related ticket. Feel free to assign to me: https://github.com/pydantic/pydantic-core/issues/358

File: docs/examples/types_import_string_1.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+# You can use the imported function as you would expect
+cos_of_0 = my_cos.obj(0)
+assert cos_of_0 == 1
+
+
+# However, the imported "object" is not available outside the class
+try:
+    math.cos(0)
+except (NameError, TypeError) as e:
+    print(e)
```
[the remark this answers is not in the document]
Comment (ybressler) [reply docs/examples/types_import_string_1.py:20]: Removed.

File: docs/examples/types_import_string_3.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ybressler) [reply docs/examples/types_import_string_3.py:36]: Let's create a new issue and close this one.

[the remark this answers is not in the document]
Comment (ybressler) [reply docs/examples/types_import_string_3.py]: Ticket: https://github.com/pydantic/pydantic/issues/4958

**Comment by ybressler** [comment]
Addressed concerns in PR. Also, created 2 follow up issues regarding concerns. Let me know what further changes are required.

**Comment by ybressler** [comment]
Waiting on review from you @samuelcolvin + @hramezani

**Pull request merged**